sql/events.go

```diff
@@ -0,0 +1,239 @@
+// need to be preserved as defined and evaluated in the first place.
+func (e *EventDetails) GenerateCreateEventStatement() string {
+	stmt := fmt.Sprintf("CREATE")
+	if e.Definer != "" {
+		stmt = fmt.Sprintf("%s DEFINER = %s", stmt, e.Definer)
+	}
+	stmt = fmt.Sprintf("%s EVENT `%s`", stmt, e.Name)
+
+	if e.HasExecuteAt {
+		stmt = fmt.Sprintf("%s ON SCHEDULE AT '%s'", stmt, e.ExecuteAt.Format(EventTimeStampFormat))
+	} else {
+		val, field := e.ExecuteEvery.GetIntervalValAndField()
+		// STARTS should be NOT null regardless of user definition
+		stmt = fmt.Sprintf("%s ON SCHEDULE EVERY %s %s STARTS '%s'", stmt, val, field, e.Starts.Format(EventTimeStampFormat))
+		if e.HasEnds {
+			stmt = fmt.Sprintf("%s ENDS '%s'", stmt, e.Ends.Format(EventTimeStampFormat))
+		}
+	}
+
+	if e.OnCompletionPreserve {
+		stmt = fmt.Sprintf("%s ON COMPLETION PRESERVE", stmt)
+	} else {
+		stmt = fmt.Sprintf("%s ON COMPLETION NOT PRESERVE", stmt)
+	}
+
+	stmt = fmt.Sprintf("%s %s", stmt, e.Status.String())
+
+	if e.Comment != "" {
+		stmt = fmt.Sprintf("%s COMMENT '%s'", stmt, e.Comment)
+	}
+
+	stmt = fmt.Sprintf("%s DO %s", stmt, e.Definition)
+	return stmt
+}
+
+// EventStatus represents an event status that is defined for an event.
+type EventStatus byte
+
+const (
+	EventStatus_Enable EventStatus = iota
+	EventStatus_Disable
+	EventStatus_DisableOnSlove
+)
+
+// String returns the original SQL representation.
+func (e EventStatus) String() string {
+	switch e {
+	case EventStatus_Enable:
+		return "ENABLE"
+	case EventStatus_Disable:
+		return "DISABLE"
+	case EventStatus_DisableOnSlove:
+		return "DISABLE ON SLAVE"
+	default:
+		panic(fmt.Errorf("invalid event status value `%d`", byte(e)))
+	}
+}
+
+func GetEventStatusFromString(status string) (EventStatus, error) {
+	switch strings.ToLower(status) {
+	case "enable":
+		return EventStatus_Enable, nil
+	case "disable":
+		return EventStatus_Disable, nil
+	case "disable on slave":
+		return EventStatus_DisableOnSlove, nil
+	default:
+		// use disable as default to be safe
+		return EventStatus_Disable, fmt.Errorf("invalid event status value: `%s`", status)
+	}
+}
+
+// EventOnScheduleEveryInterval is used to store ON SCHEDULE EVERY clause's interval definition.
+// It is equivalent of expression.TimeDelta without microseconds field.
+type EventOnScheduleEveryInterval struct {
+	Years   int64
+	Months  int64
+	Days    int64
+	Hours   int64
+	Minutes int64
+	Seconds int64
+}
+
+func NewEveryInterval(y, mo, d, h, mi, s int64) *EventOnScheduleEveryInterval {
+	return &EventOnScheduleEveryInterval{
+		Years:   y,
+		Months:  mo,
+		Days:    d,
+		Hours:   h,
+		Minutes: mi,
+		Seconds: s,
+	}
+}
+
+// GetIntervalValAndField returns ON SCHEDULE EVERY clause's
+// interval value and field type in string format
+// (e.g. returns "'1:2'" and "MONTH_DAY" for 1 month and 2 day
+//
+//	or returns "4" and "HOUR" for 4 hour intervals).
+func (e *EventOnScheduleEveryInterval) GetIntervalValAndField() (string, string) {
```

Comment:
Generally GMS uses golang libraries for this kind of result. For example, why not a `time.Duration?`

Reply:
Also, after above change call this `Interval`

Reply:
The main reason for using `EventOnScheduleEveryInterval` is to be able to get the interval value as how user defined when creating the event the first time such as `1:2 HOUR_MINUTE` whereas if we store it as `time.Duration`, we won't be able to get the exact interval that the user-defined, which is used for constructing `CREATE EVENT` statement or populating interval values in `SHOW EVENTS` or `information_schema.events` tables.

Reply:
I see. I agree a Duration isn't really usable here.

Consider moving this logic into a normal function in the plan package where it's ultimately used.